common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/microsoftsts/MicrosoftStsOAuth2Strategy.java

```diff
@@ -596,6 +597,40 @@ private void validateAuthScheme(@NonNull final MicrosoftStsTokenRequest request,
         }
     }
 
+    /**
+     * Validates that the token response contains an access token, id_token and refresh token
+     *
+     *      * @param response The idp response.
+     * @throws ClientException
+     */
+    private void validateTokensAreInResponse(@NonNull final MicrosoftStsTokenResponse response)
+            throws ClientException {
+
+        String clientException = null;
+        String description = null;
+
+        if(response.getIdToken() == null){
+            clientException = ClientException.ID_TOKEN_MISSING;
+            description = "id token (required) is missing from token response.";
+        }
+
+        if(response.getAccessToken() == null){
+            clientException = ClientException.ACCESS_TOKEN_MISSING;
+            description = "access token (required) is missing from token response.";
+        }
+
+        if(response.getRefreshToken() == null){
+            clientException = ClientException.REFRESH_TOKEN_MISSING;
+            description = "refresh token (required) is missing from token response.";
+        }
```

Comment:
Or... optionally concat the error strings together before throwing: would contain the entirety of the error in that case

Reply:
Yes.  I was aware of that.  Since all of these are required another option would be to have a single exception.  Token response is missing expected token of type: ""  ... i'm leaning that way.